Move fee estimator into validationinterface/cscheduler thread

TheBlueMatt opened the pull request:
This does a few things in a PR, so I'm happy to split it up if people prefer.

a) It clarifies the validationinterface stuff into a mempoolinterface and the validationinterface, splitting the two up into separate systems with the mempool creating events for mempoolinterface and validationinterface functions coming from validation. They are conjoined in the backend to provide ordering guarantees for listeners which are *both* mempool interfaces and validation interfaces, but are otherwise fully separated. There are a few cleanups that are enabled here, which I went ahead and did.

b) fee estimator becomes a listener to the mempool interface which provides a bit of additional info about transactions being added more than just the tx (eg feerate, etc)

c) a few minor edge cases in the fee estimator are handled (witness malleation changing feerate, and handling reorgs better, specifically).

TheBlueMatt committed:
Clarify validationinterface notification ordering

TheBlueMatt committed:
Pass MemPoolRemovalReason out through TransactionRemovedFromMempool

TheBlueMatt committed:
Split removeRecursive into calculate/remove steps

TheBlueMatt committed:
Track the set of transactions removed/conflicted in removeForBlock

TheBlueMatt committed:
BlockConnected(vtxConflicted) -> New MempoolInterface callback

This removes the whole ConnectTrace object, which may make it
slightly harder to remove the unbounded-memory-during-reorg bug
by throwing blocks out of memory and re-loading them from disk
later. Comments are added to validationinterface to note where
this should likely happen instead of ConnectTrace.

TheBlueMatt committed:
Pass new block height through MempoolUpdatedForBlock

TheBlueMatt committed:
Track witness hash malleation in fee estimator

TheBlueMatt committed:
Handle reorgs much better in fee estimator.

There are lots of better things we can do, but this is much better
than it was.

TheBlueMatt committed:
Make fee estimation not need CTxMemPoolEntry is block processing

This makes the processBlock function in CBlockPolicyEstimator much
more simlar to MempoolInterface's MempoolUpdatedForBlockConnect
callback.

TheBlueMatt commented:
Obviously this cleans up some of the interface gook from #10286, and adds some more docs that probably could have been there. Next up after this one is moving alll the remaining events all into the background thread, with associated net_processing cleanups in the process.

TheBlueMatt commented:
Rebased.

theuni commented:
Concept ACK.

Though I'm nervous about backgrounding more stuff before safeguarding against outside (unsynchronized) access. For example, the mempool/feeestimator may now be significantly ahead of what the wallet has processed. So calling CreateTransaction when we're in the middle of connecting 10 blocks seems... dangerous, even with the BlockUntilSyncedToCurrentChain hack.

TheBlueMatt commented:
Hmm? If anything the fee estimator should be at much lower risk there than for wallet...at worst you'll get out of date fee estimates...something you'd get anyway because you're calling the fee estimator in the middle of syncing stuff. Note that #11824 also keeps the queue depth at no more than 10, though I dont really want to rely on that and would prefer to implement the dump/reload-from-disk proposal suggested at https://github.com/bitcoin/bitcoin/pull/11775/files#diff-e8d9e22d9683f73a9fb8399be0dab640R145

TheBlueMatt committed:
Add a parallel validation interface for mempool events.

Because many listeners (eg wallet) will want both types of events
to be well-ordered, they are parallel and connected on the backend.
However, they are exposed separately to clients to separate the
logic (and because, hopefully, eventually, they can be exposed to
external clients of Bitcoin Core via libconsensus or similar).

TheBlueMatt committed:
Process fee estimates in the background via MempoolInterface

TheBlueMatt commented:
Rebased.

jimpo commented on src/policy/fees.cpp:
```diff
@@ -580,13 +581,20 @@ void CBlockPolicyEstimator::TransactionRemovedFromMempool(const CTransactionRef
         return false;
     }
+    if (pos->second.witness_hash != tx->GetWitnessHash()) {
+        // In the case that we saw a tx which had its witness malleated,
```
commit: Track witness hash malleation in fee estimator

Why don't we want to record the blocks to confirm in this case?

TheBlueMatt replied:
Because we don't know how long its been around/it waited.

jimpo replied:
I don't understand. We have the record of when the tx entered the mempool (albeit with different witness data). How does the malleated witness change things?

TheBlueMatt replied:
Because clearly there was a propagation issue. We dont know whether the lower-feerate version (or some other version) was the one most miners saw or the higher-feerate version. Better to just ignore it.

jimpo commented on src/test/policyestimator_tests.cpp:
```diff
@@ -77,6 +84,13 @@ BOOST_AUTO_TEST_CASE(BlockPolicyEstimates)
         block.clear();
         // Check after just a few txs that combining buckets works as expected
         if (blocknum == 3) {
+            // Wait for fee estimator to catch up
```
commit: Process fee estimates in the background via MempoolInterface

This code block crops up a lot. Move to a new function `std::future<void> ValidationInterfaceQueueClear()` (or some much better name that returns a future) in `validationinterface.h`.

TheBlueMatt replied:
This was written before that existed. Fixed.

jimpo commented on src/policy/fees.h:
```diff
@@ -196,20 +197,18 @@ class CBlockPolicyEstimator
-    /** Remove a transaction from the mempool tracking stats*/
-    bool removeTx(uint256 hash);
+    // MempoolInterface functions:
+    void TransactionAddedToMempool(const NewMempoolTransactionInfo& info, const std::vector<CTransactionRef>& txn_replaced) override;
```
commit: Process fee estimates in the background via MempoolInterface

Make the callbacks protected?

TheBlueMatt replied:
Fixed.

jimpo reviewed (commented):
Concept ACK. Overall, this looks awesome and simplifies a lot of code. I'm all for decoupling stuff through the ValidationInterface.

I did find it hard to review because it is so large and would prefer it be broken up into fee estimate stuff and non-fee estimate specific refactors.

Also, there's a typo in the commit title of "Make fee estimation not need CTxMemPoolEntry is block processing"

TheBlueMatt committed:
Remove boost::signals from txmempool, call GetMainSignals() directly

TheBlueMatt committed:
Add txn_replaced to Added callback, remove lReplaced ATMP arg

TheBlueMatt commented on src/policy/fees.h:
```diff
@@ -185,20 +186,30 @@ class CBlockPolicyEstimator
+    };
+
+    /** Remove a transaction from the mempool tracking stats*/
+    void removeTx(std::map<uint256, TxStatsInfo>::iterator pos, bool inBlock);
```
Happy to rename it, but honestly I dont remember what the different responsibilities are. Its been so long since I wrote this I dont know how it works anymore. Any suggestions?

jimpo replied:
Actually, it's fine leaving this one as is because in the later commit (Process fee estimates in the background via MempoolInterface), the other `removeTx` gets renamed to `removeTxNotInBlock`.

Not really necessary, but maybe rename the other function to `removeTxNotInBlock` in this commit instead of a later commit to avoid the confusion here.

TheBlueMatt commented on src/policy/fees.cpp:
```diff
@@ -577,43 +582,45 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
-    feeStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
-    shortStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
-    longStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
+    mapMemPoolTxs.erase(pos);
```
Would also need to copy blockHeight as well, at which point you'd just be copying the whole struct so that you can remove it.

jimpo replied:
Yeah, I think copying the struct for the sake of simplifying the control flow (by only erasing in one place) is actually good. Also would just make it more readable to have a descriptive variable name like `tx_info` instead of `pos->second`.

TheBlueMatt commented on src/policy/fees.h:
```diff
@@ -185,20 +186,30 @@ class CBlockPolicyEstimator
      */
     static constexpr double FEE_SPACING = 1.05;
 
+    struct TxStatsInfo
```
Probably just a change in code between when I wrote this and now.

TheBlueMatt commented on src/policy/fees.cpp:
```diff
@@ -504,20 +505,24 @@ void TxConfirmStats::removeTx(unsigned int entryHeight, unsigned int nBestSeenHe
     }
 }
 
+void CBlockPolicyEstimator::removeTx(std::map<uint256, TxStatsInfo>::iterator pos, bool inBlock) {
```
Fixed.

TheBlueMatt commented on src/txmempool.cpp:
```diff
@@ -443,7 +443,8 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
     nTransactionsUpdated++;
-    if (minerPolicyEstimator) {minerPolicyEstimator->removeTx(hash, false);}
+    if (reason != MemPoolRemovalReason::BLOCK && minerPolicyEstimator) {
+        minerPolicyEstimator->removeTx(hash);}
```
Fixed.

TheBlueMatt commented on src/txmempool.cpp:
```diff
@@ -599,6 +589,7 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef>& vtx, unsigne
         removeConflicts(*tx, txn_conflicts);
         ClearPrioritisation(tx->GetHash());
     }
+    if (minerPolicyEstimator) {minerPolicyEstimator->processBlock(nBlockHeight, txn_removed_in_block);}
```
Fixed.

TheBlueMatt commented on src/validationinterface.h:
```diff
@@ -70,6 +70,28 @@ void SyncWithValidationInterfaceQueue();
+     * Information about a newly-added-to-mempool transaction.
+     */
+    struct NewMempoolTransactionInfo {
+        //!A shared pointer to the transaction which was added.
```
Fixed.

TheBlueMatt commented on src/validationinterface.h:
```diff
@@ -70,6 +70,28 @@ void SyncWithValidationInterfaceQueue();
+    struct NewMempoolTransactionInfo {
+        //!A shared pointer to the transaction which was added.
+        CTransactionRef m_tx;
+        //! The fee the added thransaction paid
```
Fixed.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -152,9 +152,9 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-void CMainSignals::TransactionAddedToMempool(const CTransactionRef &ptx) {
-    m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
-        m_internals->TransactionAddedToMempool(ptx);
+void CMainSignals::TransactionAddedToMempool(const CTransactionRef &ptx, const std::shared_ptr<std::vector<CTransactionRef>>& txn_replaced) {
```
Obviousy avoiding the move is a bit more performant, hence its use here. The use of the move in MempoolUpdatedForBlockConnect was to reduce diff size.

TheBlueMatt commented on src/txmempool.cpp:
```diff
@@ -418,8 +418,10 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
+    if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT &&
+        reason != MemPoolRemovalReason::REPLACED) {
         // BLOCK and CONFLICT callbacks are generated in removeForBlock
+        // REPLACED txn are included in TransactionAddedToMempool from ATMP
```
Done.

TheBlueMatt commented on src/init.cpp:
```diff
@@ -197,7 +197,8 @@ void Shutdown()
     if (peerLogic) UnregisterValidationInterface(peerLogic.get());
-    if (g_connman) g_connman->Stop();
+    if (peerLogic) UnregisterMempoolInterface(peerLogic.get());
+    if(g_connman) g_connman->Stop();
```
Fixed.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1226,7 +1226,16 @@ void CWallet::TransactionRemovedFromMempool(const CTransactionRef &ptx, MemPoolR
+        SyncTransaction(ptx);
+        TransactionRemovedFromMempool(ptx, MemPoolRemovalReason::CONFLICT);
+    }
+    // We can ignore tx_removed_in_block, we'll get it in BlockConnected
```
Because we still need it in BlockConnected - there may be txn that weren't in our mempool that got confirmed, so copying it would be useless.

TheBlueMatt commented on src/txmempool.cpp:
```diff
@@ -470,11 +470,8 @@ void CTxMemPool::CalculateDescendants(txiter entryit, setEntries &setDescendants
 }
 
-void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReason reason)
+void CTxMemPool::calculateRemoveRecursive(const CTransaction &origTx, setEntries &setAllRemoves)
```
Done (bonus: also fixed the whitespace issues in that commit).

TheBlueMatt commented on src/txmempool.h:
```diff
@@ -492,6 +492,8 @@ class CTxMemPool
 
     std::vector<indexed_transaction_set::const_iterator> GetSortedDepthAndScore() const;
 
+    void calculateRemoveRecursive(const CTransaction &tx, setEntries &stage);
```
This PR was created long before the introduction of that part of the style guide, so it matches the capitalization of removeRecursive.

sdaftuar commented:
I've started reviewing and testing the first few commits in this PR, but I already agree with @jimpo above, that this would be easier to review if it was split up.

Perhaps we could first have a PR that changed the callback interface around, and then another PR that reworks the fee estimator to take advantage of the new interface?

(also this needs rebase)

ajtowns commented:
ConceptACK for what it's worth, but definitely split this up -- it overflows my mental stack, and both times I've tried to review it it's already had conflicts with master that make a review not crazy useful...

Breaking out the trivial bits, at least "Clarify validationinterface notification ordering" and "Remove useless scoped (sic?) in AcceptToMemoryPool" seems like a good first move. Maybe "Split removeRecursive" as well. Otherwise just splitting the remaining patches into mempool/validationinterface vs fee-estimation focused seems plausible.

MarcoFalke commented:
@bugthebluematt ^

TheBlueMatt committed:
Remove useless scope in AcceptToMemoryPoolWorker

TheBlueMatt commented:
Now based on #12979.
